src/process.ml

```diff
@@ -152,18 +164,32 @@ let spawn
 
   let redirections, redirection_count = build_redirection_array redirect in
 
+  let gc_arg_handles, c_args = c_string_array arguments in 
+  let gc_env_handles, c_env = c_string_array env in
+
+  let c_cwd = cwd |> Ctypes.(CArray.of_string) |> Ctypes.CArray.start in
+  let c_path = path |> Ctypes.(CArray.of_string) |> Ctypes.CArray.start in
+
+  (*
+    Ensure that the GC maintains handles to all the strings we're passing through the duration of the call.
+
+    This is important because the runtime is released, and there is a potential race condition if the
+    GC is triggered between releasing the runtime and calling [uv_spawn].
+  *)
+  let gc_handles = ref (Some (gc_arg_handles, gc_env_handles, c_cwd, c_path)) in
```

Comment:
This method of retaining references will break in multithreaded code. Luv uses another method, which is to refer to a value locally further down in the function by passing it to `Sys.opaque_identity`, like so:

https://github.com/aantron/luv/blob/59a5418145bc6af3570ec5341d108f8df4e4da35/src/stream.ml#L164-L173

In this case, `buffers` and `iovecs` must be retained by the GC across the call to `try_write`.

Can you convert your code to use `Sys.opaque_identity`, and see if that also works for you? You should be able to also undo the changes to `luv_spawn`'s Ctypes "signature."

Reply:
Now that I looked at the code in an editor, this is fine &mdash; it's a local ref. I had mistaken it for a global one.